### PR TITLE
add `MutUntyped::map_unchanged`

### PR DESCRIPTION
**Pull request by Serverator**
### **Adopted #6430**

# Objective

`MutUntyped` is the untyped variant of `Mut<T>` that stores a `PtrMut` instead of a `&mut T`. Working with a `MutUntyped` is a bit annoying, because as soon you want to use the ptr e.g. as a `&mut dyn Reflect` you cannot use a type like `Mut<dyn Reflect>` but instead need to carry around a `&mut dyn Reflect` and a `impl FnMut()` to mark the value as changed.
## Solution

    * Provide a method `map_unchanged` to turn a `MutUntyped` into a `Mut<T>` by mapping the `PtrMut<'a>` to a `&'a mut T`
      This can be used like this:

```rust
// SAFETY: ptr is of type `u8`
let val: Mut<u8> = mut_untyped.map_unchanged(|ptr| unsafe { ptr.deref_mut::<u8>() });

// SAFETY: from the context it is known that `ReflectFromPtr` was made for the type of the `MutUntyped`
let val: Mut<dyn Reflect> = mut_untyped.map_unchanged(|ptr| unsafe { reflect_from_ptr.as_reflect_ptr_mut(ptr) });
```

Note that nothing prevents you from doing

```rust
mut_untyped.map_unchanged(|ptr| &mut ());
```

or using any other mutable reference you can get, but IMO that is fine since that will only result in a `Mut` that will dereference to that value and mark the original value as changed. The lifetimes here prevent anything bad from happening.
## Alternatives

1. Make `Ticks` public and provide a method to get construct a `Mut` from `Ticks` and `&mut T`. More powerful and more easy to misuse.
2. Do nothing. People can still do everything they want, but they need to pass (`&mut dyn Reflect, impl FnMut() + '_)` around instead of `Mut<dyn Reflect>`

## Changelog

- add `MutUntyped::map_unchanged` to turn a `MutUntyped` into its typed counterpart

### CODE REVIEW

File: crates/bevy_ecs/src/change_detection.rs [hunk 1]
```diff
@@ -735,6 +735,34 @@ impl<'a> MutUntyped<'a> {
         self.value.as_ref()
     }
 
+    /// Turn this [`MutUntyped`] into a [`Mut`] by mapping the inner [`PtrMut`] to another value,
+    /// without flagging a change. [`MutUntyped`] equivalent of [`Mut::map_unchanged`].
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/change_detection.rs:739]:
```suggestion
    /// without flagging a change. 
    /// This function is the untyped equivalent of [`Mut::map_unchanged`].
```

This phrasing reads better IMO.

File: crates/bevy_ecs/src/change_detection.rs [hunk 2]
```diff
@@ -1034,4 +1063,40 @@ mod tests {
...
+        let mut value: i32 = 5;
+        let value = MutUntyped {
+            value: PtrMut::from(&mut value),
+            ticks,
+        };
+
+        let reflect_from_ptr = <ReflectFromPtr as FromType<i32>>::from_type();
+
+        let mut new = value.map_unchanged(|ptr| {
+            // SAFETY: ptr has type of ReflectFromPtr
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/change_detection.rs:1091]:
```suggestion
            // SAFETY: The underlying type of `ptr` matches `reflect_from_ptr`.
```

**Review by JoJoJet — APPROVED** [review]
Thank you for adopting this!

**Review by alice-i-cecile — APPROVED** [review]
I agree with the doc suggestions but won't block on them.

**Comment by Serverator** [comment]
Uh huh... For some reason CI is failing now, but I don't really understand why, I only commited the doc changes

**Comment by JoJoJet** [comment]
One of my suggestions must have had an extra space or something -- you should be able to just run `cargo fmt` to fix it.

File: crates/bevy_ecs/src/change_detection.rs [hunk 3]
```diff
@@ -735,6 +735,35 @@ impl<'a> MutUntyped<'a> {
         self.value.as_ref()
     }
 
+    /// Turn this [`MutUntyped`] into a [`Mut`] by mapping the inner [`PtrMut`] to another value,
+    /// without flagging a change.
```
Comment (JoJoJet) [inline crates/bevy_ecs/src/change_detection.rs:739]:
```suggestion
    /// without flagging a change.
```

Comment (Serverator) [reply crates/bevy_ecs/src/change_detection.rs:739]: Let's try...

**Comment by mockersf** [comment]
Miri failure is due to a but in rust nightly, it should be fixed tomorrow

**Review by mockersf — APPROVED** [review]

**Pull request merged**